drivers: display: st7789v: implement display_set_orientation API

LuskeyNoah opened the pull request:
Add support for 135x240, 240x320, 240x240 displays at 0, 90, 180, and 270 degree orientation.

closes #32286

danieldegrasse commented on drivers/display/display_st7789v.c:
```diff
@@ -60,23 +61,22 @@ struct st7789v_data {
 
-	struct spi_buf tx_buf = { .buf = &cmd, .len = 1 };
-	struct spi_buf_set tx_bufs = { .buffers = &tx_buf, .count = 1 };
+	struct spi_buf tx_buf = {.buf = &cmd, .len = 1};
```
Did you run `clang-format` on this driver file? I have no issue with these sorts of cleanups, but please put the clang-format changes into a separate commit- otherwise this sort of PR is difficult to review, and the git blame will be confusing as well :)

LuskeyNoah replied:
Yes ran clang-format. Should've turned off format on save 😨 

For separate commits - do you want it as completely separate PRs or two commits (1 for reformat, 1 for the actual enhancement) in the same PR?

danieldegrasse replied:
Two in one PR is totally okay- I'd just like separate commits to make things easier to follow in the git history :)

LuskeyNoah replied:
Fixed!

Commit:
drivers: display: st7789v: run clang-format

Run clang-format on st7789v driver implementation

Signed-off-by: Noah Luskey <noah@vvvvvvvvvv.io>

Commit:
drivers: display: st7789v: implement display_set_orientation API

Add support for communicating to 135x240, 240x320, 240x240 displays
at 0, 90, 180, and 270 degree orientation.

Signed-off-by: Noah Luskey <noah@vvvvvvvvvv.io>

Snevzor commented:
Hello @LuskeyNoah, I'm also working on a similar PR for the ssd16xx driver. I'm curious why you chose not to switch x/y_resolution in the get_capabilities function, wich is done here: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/display/display_ili9xxx.c#L246.

I was traveling the ili9xxx road and got into an issue by switching those parameters in the LVGL glue code here: https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/lvgl/lvgl.c#L107. The allocation function assumes X matches the display width and Y matches the the display height, which is not the case anymore when the display driver already has them switched. As a result, the lvgl_init fails for non-rectangular resolutions.

I decided that the driver does not need to switch the resolutions (like with your implementation), but then LVGL needs to do it, so I think the glue code needs a "disp_drv.rotated = ..." (in lvgl_init()) based on the current orientation from the display capabilities. This is what I'm going for at the moment but I wonder if touching the glue code like that will break other display orientation implementations.

Also, based on my findings I actually think that ili9xxx + lvgl sample does not work currently when the orientation is 90/270 degrees, but I cannot verify because I lack the necessary HW to test.

What are your thoughts on this?

LuskeyNoah commented:
>  I'm curious why you chose not to switch x/y_resolution in the get_capabilities function, wich is done here

@Snevzor - good catch. This is an oversight on my part. I'm pushing an update, as it looks like [other drivers](https://github.com/zephyrproject-rtos/zephyr/blob/598ab6b39a5e38466a286fb317f39bcd34b8ad20/drivers/display/display_gc9x01x.c#L636) choose to take the approach of reporting the swapped dimensions.

LuskeyNoah committed:
Merge branch 'zephyrproject-rtos:main' into main

Snevzor commented:
@LuskeyNoah I'm still a bit confused about the correct approach on this. Please refer to my comment here: https://github.com/zephyrproject-rtos/zephyr/pull/73360#issuecomment-2143410580.

LuskeyNoah commented:
It's becoming clear I do not have the proper hardware to fully test this. Closing. @Snevzor please reopen and/or contribute to it if desired. I will reopen if I can fully test this.

LuskeyNoah commented:
It's becoming clear I do not have the proper hardware to fully test this. Closing. @Snevzor please reopen and/or contribute to it if desired. I will reopen if I can fully test this.

Snevzor commented:
@LuskeyNoah I unfortunately also do not have a st7789 based display laying around, I'm sorry.

DBS06 commented:
Hi, because I opened a PR for porting the Adafruit Feather ESP32-S2 (https://github.com/zephyrproject-rtos/zephyr/pull/85405) and I have a Adafruit Feather ESP32-S2 TFT which has a ST7789v and a "1.14" IPS TFT with 240x135 pixels" (actually the pixels are 135x240).
I tested the code with the `samples/subsys/display/lvgl` sample, the orientation change works in principal.

My first approach was this:
```
	int err = display_set_orientation(display_dev, DISPLAY_ORIENTATION_ROTATED_270);
	if (err == 0) {
		LOG_INF("Display orientation set to rotated 270 degrees");
	} else {
		LOG_ERR("Failed to set display orientation");
	}
```
The rotation works in principal but it gives me this result:
<img width="617" alt="Screenshot 2025-02-12 130615" src="https://github.com/user-attachments/assets/2a514cde-46bf-46b0-94e5-3eebf7d8849e" />

This looks like lvgl does not recognize a change in orientation from the display-driver-module. So I tried the following: set `CONFIG_LV_Z_AUTO_INIT=n` and called `lvgl_init()` after changing the orientation, but actually this had no effect at all.

After that I tried to just use `lv_display_set_rotation(lv_display_get_default(), LV_DISPLAY_ROTATION_270);` and had the following weird result:
<img width="745" alt="Screenshot 2025-02-12 131358" src="https://github.com/user-attachments/assets/3675493d-fdc7-41e9-99f1-0f4ab232ebf9" />

By doing both `display_set_orientation()` and `lv_display_set_rotation()`, unset `CONFIG_LV_Z_AUTO_INIT=n` and not calling `lvgl_init()` manually, see code-snippet here:
```
	int err = display_set_orientation(display_dev, DISPLAY_ORIENTATION_ROTATED_270);
	if (err == 0) {
		LOG_INF("Display orientation set to rotated 270 degrees");
	} else {
		LOG_ERR("Failed to set display orientation");
	}
	lv_display_set_rotation(lv_display_get_default(), LV_DISPLAY_ROTATION_270);
```
It works as it should:
<img width="904" alt="Screenshot 2025-02-12 131433" src="https://github.com/user-attachments/assets/a217dc88-478d-4d37-a95a-94b7d2270937" />

My branch can be viewed here:
https://github.com/DBS06/zephyr/tree/st7789v_implement_display_set_orientation_API

If this behavior is acceptable I would reopen this PR.

Snevzor commented:
Hi @DBS06,

For this to work well with LVGL this is needed: https://github.com/zephyrproject-rtos/zephyr/pull/73556.
Also, in context to why this PR was abandoned, https://github.com/zephyrproject-rtos/zephyr/issues/82582 needs to be clarified fist.

Maybe check with @danieldegrasse?

Br,
Sven

DBS06 commented:
@Snevzor Thanks for the fast response and clarification!
